README.md
# PlateLayout

## Installation

You can install the package from github with:

``` r
# install.packages("devtools")
devtools::install_github("crukci-bioinformatics/PlateLayout")
```

## Introduction

This package takes an experimental design table and generates a
"pseudo-block-randomised" plate layout for it. 

The package optimises the layout of the plate, and then randomises the columns
and rows. The randomisation process is carried out multiple times and the
distance between replicate samples is used to score each randomistation to see
how well distributed/badly clumped each sample group is. The "best"
randomistation is then delivered.

The script uses one column, defined by the `primaryGroup` argument (by default
"SampleGroup"), for the optimisation/randomistaion process. If there are known
batches or other factors of interest in the experiment (e.g. different
extraction dates, sex etc.) the optimisation part of script does not consider
these, however, they can be considered in the scoring of each random layout via
the `batchColumns` argument.

## Inputs

The input table should be a tab separated file with samples in rows and
experimental factors and any other metadata in columns.

## Outputs

The outputs are one image (png) file for each experimental factor considered in
the layout and a table with the layout.

## Usage

### R functions

##### Randomisation

The main randomisation command is `randomizeSinglePlate`. The required input is
the metadata table as a data.frame or tibble.

By default it uses the "SampleGroup" column to optimise the layout,
however, any other column can be used instead by specifying the argument
`primaryGroup`, and in fact there are no required column headers.

Additional columns can be passed in the `batchColumns` argument, these will be
used to select the best layout from `nIter` random layouts.

The default return value is a randomised table. You can then use `plotPlate` to
plot the distribution of variables contained in the columns of the metadata 
table across the plate. e.g.:

```
library(tidyverse)
library(PlateLayout)

designSheet <- system.file("extdata", 
                           "metadata_5x3.tsv",
                           package = "PlateLayout")
bColumns <- c("ExtractionInformation", "PassageNumber") 

plateLayout <- read_tsv(designSheet) %>%  
               mutate(Replicate = str_extract(SampleName, "[0-9]$")) %>%  
               mutate(SampleGroup = str_remove(SampleName, "_[0-9]$")) %>%  
               randomizeSinglePlate(batchColumns = bColumns, nIter = 100) 

pdf("Plate_Layout.5x3.pdf")
bColumns %>%  
    map(~plotPlate(plateLayout, plotCol = .x))
dev.off()

plateLayout %>%  
    select(-SampleGroup, -RowID) %>%  
    write_tsv("Plate_Layout.5x3.tsv")
```

Alternatively, you can pass `randomizeSinglePlate` an output file prefix and it
will write out the randomized plate layout as a tsv file and individual plots
for each of the `batchColumns`.

```
designSheet <- system.file("extdata", 
                           "metadata_12x3.tsv",
                           package = "PlateLayout")
bColumns <- c("ExtractionInformation", "PassageNumber") 
outputFile <- "Plate_Layout.12x3"

read_tsv(designSheet) %>%  
    randomizeSinglePlate(outputFile = outputFile, 
                         batchColumns = bColumns, 
                         nIter = 100) 
```

**Note:** by default the script runs in parallel across 4 cores. You can change
this with the `nCores` argument.

#### CRUK CI Bioinformatics metadata template

If you are using a CRUK CI Bioinformatics Core metadata template Excel file, then
there is a command `loadMetaForm` that will load a and tidy it into a table suitable for 
use with the above randomisation function. It will fix the column headers and remove 
empty columns. 

### Bash script

In the *bin* directory of the Git repository is a bash script
`RandomizeSinglePlate.sh` that can be used at the command line to run the
randomisation. e.g.

```
RandomizeSinglePlate.sh -d example_metadata.tsv \
                        -o Plate_Layout \
                        -b SampleType,Patient \
                        -r 100
```

Please use `RandomizeSinglePlate.sh -H` for usage.

## Plate Layout Process

The algorithm for randomising the plate layout is as follows:

### Optimisation
1. Randomise the order of the sample groups and the order of the replicates
   within each sample group:  
    a) assign each sample group a random number from 1-nGroups  
    b) assign each sample a random number from 1-nSamples  
    c) sort the table by group and then sample  

2. Assign each sample to a well by traversing the plate diagonally:    
    a) determine the number of columns that will be needed to accomodate all the
samples on the plate    
    b) the first sample in the table is then assigned to A1, the second to B2
      and so on. On reaching the final column continue the next row from column
      1, and continue the next column from Row A when Row H has been reached.  
    The purpose of assigning wells in this way is to distribute the
      replicates from each sample group across the rows and the columns so
      that, as much as possible, replicates are not in the same row or column as
      others from the same sample group.  

### Randomisation

5. Randomise rows and columns as blocks - this maintains the distribution of
   replicates relative to each other, but randomises the overall plate layout.  
    a) each row is assigned a randomn number from 1-8 and the rows are re-ordered
     according to this  
    b) each column is assigned a randomn number from 1-nColumns and the columns
      are ordered according to this  

    e.g. for 5 rows with 40 samples:
    
        Step 4 ->
             1  17  33   9  25
            26   2  18  34  10
            11  27   3  19  35
            36  12  28   4  20
            21  37  13  29   5
             6  22  38  14  30
            31   7  23  39  15
            16  32   8  24  40

        Step 5i ->
            31   7  23  39  15
            11  27   3  19  35
            21  37  13  29   5
            16  32   8  24  40
            36  12  28   4  20
             1  17  33   9  25
            26   2  18  34  10
             6  22  38  14  30

        Step 5ii ->
            39  31  23  15   7
            19  11   3  35  27
            29  21  13   5  37
            24  16   8  40  32
             4  36  28  20  12
             9   1  33  25  17
            34  26  18  10   2
            14   6  38  30  22
            
6. Score the randomised plate for distribution of each factor:  
    A score for each sample is calculated as the sum of 1 / (Euclidean Distance
    to other samples of the same type where dist is <= 2).  
    The distribution score for the plate is the sum of scores for all samples
    for all "batch columns".
                    
7. The randomisation process is carried out 10000 times and the layout with the
   lowest "distribution score" is accepted.


NEWS
VERSION 0.3.3
------------------------
Add option to switch off warning message about the RNG and future
Modify slice for new argument syntax
VERSION 0.3.2
------------------------
Update `multiprocess` to `multicore`
VERSION 0.3.1
------------------------
Bug fix - errant ':x' in the plate size check
VERSION 0.3
------------------------
Add function for loading the CRUK CI Bioinformatics metadata template and
transforming it into a suitable input table.
VERSION 0.2
------------------------
Changed the input of randomizeSinglePlate to a table rather than a file
Changed outputFile to optional - will return table as R object if no outputFile
Export the plotting function
Moved the bash script out of the R package
Renamed example data in extdata
VERSION 0.1
------------------------
Initial version - just one function: randomizeSinglePlate 


test/TestScript.sh
#!/bin/bash

# locate test script directory

SOURCE="${BASH_SOURCE[0]}"
while [ -h "$SOURCE" ]; do # resolve $SOURCE until the file is no longer a symlink
  DIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
  SOURCE="$(readlink "$SOURCE")"
  [[ $SOURCE != /* ]] && SOURCE="$DIR/$SOURCE" # if $SOURCE was a relative symlink, we need to resolve it relative to the path where the symlink file was located
done
TestDIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
TestDIR=`readlink -f $TestDIR`


# export binary file location
BinaryDIR=${TestDIR/test/bin}
export PATH=$BinaryDIR:$PATH

# tests
echo "Run: metadata_template_TEST_11x4.csv"
PlateLayoutRandomistation -d $TestDIR/metadata_template_TEST_11x4.csv -o MyPlateLayout1 -b Sex,ExtractionInformation -r 3

echo "Run: metadata_template_TEST_12-12-14.csv"
PlateLayoutRandomistation -d $TestDIR/metadata_template_TEST_12-12-14.csv -o MyPlateLayout -b Sex,ExtractionInformation,PassageNumber

echo "Run: metadata_template_TEST_12x3.csv"
PlateLayoutRandomistation -d $TestDIR/metadata_template_TEST_12x3.csv -o MyPlateLayout2 -b Sex,ExtractionInformation -r 2

echo "Run: metadata_template_TEST_12x4.csv"
PlateLayoutRandomistation -d $TestDIR/metadata_template_TEST_12x4.csv

echo "Run: metadata_template_TEST_12x8.csv"
PlateLayoutRandomistation -d $TestDIR/metadata_template_TEST_12x8.csv -o MyPlateLayout3 -G

echo "Run: metadata_template_TEST_21x8.csv"
PlateLayoutRandomistation -d $TestDIR/metadata_template_TEST_21x8.csv -b Sex,PassageNumber

echo "Run: metadata_template_TEST_23x8.csv"
PlateLayoutRandomistation -d $TestDIR/metadata_template_TEST_23x8.csv -o MyPlateLayout5 -b Sex,ExtractionInformation,PassageNumber

echo "Run: metadata_template_TEST_24x6.csv"
PlateLayoutRandomistation -d $TestDIR/metadata_template_TEST_24x6.csv -o MyPlateLayout6 -b Sex,ExtractionInformation,PassageNumber

echo "Run: metadata_template_TEST_24x7.csv"
PlateLayoutRandomistation -d $TestDIR/metadata_template_TEST_24x7.csv -b Sex,ExtractionInformation,PassageNumber -r 2

echo "Run: metadata_template_TEST_24x8.csv"
PlateLayoutRandomistation -d $TestDIR/metadata_template_TEST_24x8.csv -o MyPlateLayout8

echo "Run: metadata_template_TEST_5x3.csv"
PlateLayoutRandomistation -d $TestDIR/metadata_template_TEST_5x3.csv

echo "Run: metadata_template_TEST_100.csv"
PlateLayoutRandomistation -d $TestDIR/metadata_template_TEST_100.csv -o MyPlateLayout10 -b ExtractionInformation,PassageNumber


R/Utilities.R
# Test the distribution of a samples in a single column across the plate
# Use the sum of 1/(Euclidean distances between samples less than or equal to 2)
# For a given sample: 
#     0   0   0.5   0   0
#     0 0.707  1  0.707  0
#    0.5   1   X    1  0.5 
#     0 0.707  1  0.707  0
#     0   0   0.5   0   0

getScore <- function(rowCols){
    dst <- dist(rowCols, method = "euclidean")
    sum(1 / dst[dst <= 2])
}

calcDistrScore <- function(batchColumn, dat){
    dat %>%  
        select(bCol = batchColumn, Column, RowID) %>%  
        group_by(bCol) %>%  
        nest(dat = c(Column, RowID)) %>%  
        mutate(score = map_dbl(dat, getScore)) %>%  
        pull(score) %>%  
        sum()} 

# get distr scores for each batch column of a sample sheet 

getDistrScores <- function(samSht, batchColumns){ 
    batchColumns %>%  
        set_names(batchColumns) %>%  
        future_map_dfr(calcDistrScore, samSht)
}    

# get scores for all random layouts for all batch columns, normalised within
# each batch, sum, find best layout as minimum score

normScores <- function(scr){ scr / median(scr) }

getMinScore <- function(layoutList, batchCols){
    future_map_dfr(layoutList, getDistrScores, batchColumns = batchCols, 
                          .progress = TRUE) %>% 
        mutate_all(normScores) %>%  
        rowSums() %>%  
        which.min()
}

###############################################################################

# create a colour scheme using color brewer, few options depending on number of
# groups
getCols <- function(dat, plotCol){
    len <- pull(dat, plotCol) %>% unique() %>% length()
    if(len<=2){
        wcols <- c("#00B6ED", "#EC008C")
    }else if(len<=12){
        wcols <- RColorBrewer::brewer.pal(len, "Set3")
    }else if(len<=21){
        wcols <- c(RColorBrewer::brewer.pal(12, "Set3"), 
                   suppressWarnings(
                      RColorBrewer::brewer.pal(len - 12, "Set1")
                      ))
    }else{
        wcols <- sample(rainbow(len), len)
    }
    wcols
}

#' Plot Plate Layout
#'
#' Plot the plate with wells coloured according to a column in the samplesheet,
#' if there is column called Replicate the numbers will also be printed
#' 
#' @param dat The plate layout tables
#' @param plotCol The column of the table by which to colour the wells
#' @param wellCols A named vector of columns for colouring the wells. Names
#' should match the sample groups in plotCol
#' @return A ggplot object
#' @examples
#' library(readr)
#' designSheet <- system.file("extdata", "metadata_12x3.tsv",
#'                            package = "PlateLayout")
#' bColumns <- c("ExtractionInformation", "PassageNumber") 
#' designTable <- read_tsv(designSheet)
#' # Create an R object in the session
#' plateLayout <- randomizeSinglePlate(designTable,  
#'                                     batchColumns = bColumns,
#'                                     nIter = 100) 
#' plotPlate(plateLayout)
#' plotPlate(plateLayout, "ExtractionInformation")
#' @export
plotPlate <- function(dat, plotCol = "SampleGroup", wellCols = NULL){
    nCols <- ceiling(nrow(dat)/8)
    plt <- dat %>%  
        mutate_at(plotCol, as.character) %>%  
        ggplot(aes(x=Column, y=RowID))+
        geom_tile(aes_string(fill=plotCol), colour="black")  +
        scale_x_continuous(breaks=1:nCols) +
        scale_y_reverse(labels=LETTERS[1:8], breaks=c(1:8)) +
        theme(axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20), 
              axis.title.x=element_blank(),
              axis.title.y=element_blank()) +
        ggtitle(plotCol) +
        labs(fill=plotCol)
    if(is.null(wellCols)){ wellCols <- getCols(dat, plotCol) }
    plt <- plt + scale_fill_manual(values = wellCols)
    if(any(str_detect(colnames(dat), "[Rr]eplicate"))){
        plt <- plt + geom_text(aes(label=Replicate), size=7)
    }
    plt
}

outputPlatePlot <- function(batchColumn, datTab, outFileName){
    outFileName  <- str_c(outFileName, ".", batchColumn, ".png")
    p1 <- plotPlate(datTab, batchColumn)
    ggsave(outFileName, plot = p1)
} 

################################################################################


#' Fix metadata template
#'
#' Load and transform our CRUK CI Bioinfomratics Core metadata template into a
#' table suitable for the PlateLayout package
#' 
#' @param xlsFile The metadata template Excel file
#' @details The function takes the path for an Excel file containing completed 
#' CRUK CI metadata template and transforms it into a table suitable for use
#' with the \code{\link{randomizeSinglePlate}} function. 
#' @details When loading the Excel file the function skips the first row, 
#' making the headers from the second (hidden) row in the template. The third
#' row in the template contains a contents description for each column and so
#' is removed from the table.
#' @details Any columns that are completely empty are removed. \code{NA} is
#' replaced with a blank characer vector.
#' @return The cleaned metadata as a tibble and exports a tsv version
#' @examples
#' metadataFile <- system.file("extdata", "metadata_template_example.xls",
#'                            package = "PlateLayout")
#' metadataTable <- loadMetaForm(metadataFile)
#' metadataTable
#' @export
loadMetaForm <- function(xlsFile){
  outNam <- xlsFile %>% 
    str_replace_all(" ", "_") %>% 
    str_replace("\\.[xls]+$", ".tsv")
  message("Transforming '", xlsFile, "' to '", outNam, "'")
  dat <- read_excel(xlsFile, skip=1) %>%  
      slice(-1) %>%  
      mutate_all(as.character) %>%  
      rename(SampleGroup="Group") %>% 
      select_if(~any(!is.na(.x))) %>%  
      mutate_all(~ifelse(is.na(.x), "", .x))
  write_tsv(dat, outNam)
  dat
}



R/utilities.R

# Test the distribution of a samples in a single column across the plate
# Use the sum of 1/(Euclidean distances between samples less than or equal to 2)
# For a given sample: 
#     0   0   0.5   0   0
#     0 0.707  1  0.707  0
#    0.5   1   X    1  0.5 
#     0 0.707  1  0.707  0
#     0   0   0.5   0   0

getScore <- function(rowCols){
    dst <- dist(rowCols, method = "euclidean")
    sum(1 / dst[dst <= 2])
}

calcDistrScore <- function(batchColumn, dat){
    dat %>%  
        select(bCol = batchColumn, Column, RowID) %>%  
        group_by(bCol) %>%  
        nest(dat = c(Column, RowID)) %>%  
        mutate(score = map_dbl(dat, getScore)) %>%  
        pull(score) %>%  
        sum()} 

# get distr scores for each batch column of a sample sheet 

getDistrScores <- function(samSht, batchColumns){ 
    batchColumns %>%  
        set_names(batchColumns) %>%  
        future_map_dfr(calcDistrScore, samSht)
}    

# get scores for all random layouts for all batch columns, normalised within
# each batch, sum, find best layout as minimum score

normScores <- function(scr){ scr / median(scr) }

getMinScore <- function(layoutList, batchCols){
    future_map_dfr(layoutList, getDistrScores, batchColumns = batchCols, 
                          .progress = TRUE) %>% 
        mutate_all(normScores) %>%  
        rowSums() %>%  
        which.min()
}

###############################################################################

# plot the plate with wells coloured according to a column in the samplesheet
# if the column is the "SampleGroup" the replicate numbers are also printed
library(RColorBrewer)

plotPlate <- function(dat, plotCol, boxColCodes = NULL){
    textSize <- ifelse(plotCol=="SampleGroup", 7, 0)
    nCols <- ceiling(nrow(dat)/8)
    len <- pull(dat, plotCol) %>% unique() %>% length()
    if(is.null(boxColCodes)) boxColCodes <- brewer.pal(len, "Set3")
    ggplot(dat, aes(x=Column, y=RowID))+
        geom_tile(aes_string(fill=plotCol), colour="black")  +
        geom_text(aes(label=Replicate), size=textSize) +
        scale_x_continuous(breaks=1:nCols) +
        scale_y_reverse(labels=LETTERS[1:8], breaks=c(1:8)) +
        scale_fill_manual(values=boxColCodes) +
        theme(axis.text.x=element_text(size=20),
              axis.text.y=element_text(size=20), 
              axis.title.x=element_blank(),
              axis.title.y=element_blank()) +
        ggtitle(plotCol) +
        labs(fill=plotCol)
}

savePlots <- function(batchColumn, datTab, outFileName){
    outFileName  <- str_c(outFileName, ".", batchColumn, ".png")
    p1 <- plotPlate(datTab, batchColumn)
    ggsave(outFileName, plot = p1)
} 


R/RandomisePlate.R
# This script contains three functions to randomise samples across
# across a 96 well plate given a sample metadata table. The distribution of the samples is optimised using the contents of a single column in the table - the `primaryGroup`.
# Steps:
# 1) randomiseSampleTable - Randomise the order of different factors in the
# "primary group` and replicates within each factor in the metadata table:
#       (1) randomise the order of the sample groups in the table
#       (2) randomise the order of the replicates within each group
# 2) addWells:
#       Assign samples to wells running diagonally across the plate 
#       If the number of columns is even this causes a problem as we cycle back
#       to A1 before completing assignments. We can get around this by adding
#       a dummy column, ordering, and then removing the dummy column
#       We also need to account for empty wells in the final row if the number
#       of samples is not divisible by 8.
# 3) randomiseWells - Randomize the wells by: 
#       (1) randomize the order of the columns except the last one (may be only
#           partially full)
#       (2) randomize the order of the rows
#       (3) move any empty wells in the final column to the bottom

randomiseSampleTable <- function(dat, primaryGroup){
    # (1) randomise the order of the sample groups in the table
    # (2) randomise the order of the replicates within each group
    dat %>% 
        group_by_at(primaryGroup) %>%  
        nest()  %>%  
        ungroup() %>%  
        mutate(Ord.Gp = sample(seq(n()))) %>% 
        unnest(everything()) %>%  
        mutate(Ord.Sam = sample(seq(n()))) %>% 
        arrange(Ord.Gp, Ord.Sam) %>%  
        select(-Ord.Gp, -Ord.Sam)
}


addWells <- function(dat){
    nCols <- ceiling(nrow(dat)/8)
    nCols <- ifelse(nCols%%2==0, nCols+1, nCols) # add dummy column if necessary
    nSams <- nrow(dat)
    tibble(RowNum = rep(1:8, nCols),
           ColNum = rep(seq(nCols), 8)) %>%
           filter(ColNum <= nCols) %>% # remove dummy columns
           rowid_to_column("WellNumber") %>%
           arrange(ColNum, RowNum) %>%  
           slice(seq(nSams)) %>%  # remove empty wells in final col
           arrange(WellNumber) %>%  
           select(-WellNumber) %>%
           bind_cols(dat)
}

randomiseWells <- function(dat){ 
    # (1) Randomize columns (except last)
    dat <- dat %>%  
        group_by(ColNum) %>%  
        nest() %>%  
        ungroup() %>%  
        mutate(Column = c(sample(seq(n()-1)), n())) %>%
        unnest(cols = everything()) 
    # (2) Randomize rows
    dat <- dat %>%  
        group_by(RowNum) %>%  
        nest() %>%  
        ungroup() %>%  
        mutate(RowID = sample(seq(n()))) %>%
        arrange(RowID) %>%  
        unnest(cols = everything()) %>%  
        arrange(Column, RowID)
    # (3) Fix empty wells in final column if necessary
    nCols <- ceiling(nrow(dat)/8) # number of columns
    nFinalCol <- nrow(dat)%%8 # number of samples in the final column
    if(nFinalCol>0){
        dat <- dat %>%  
            mutate(RowID = ifelse(Column==nCols, seq(nFinalCol), RowID))
    }
    # (4) Add row letters, add well
    dat %>%  
        mutate(Row = LETTERS[RowID]) %>%  
        select(-ColNum, -RowNum) %>%  
        mutate(Well = str_c(Row, Column))
}


# main function
plateRandomisation <- function(samSht, primaryGroup){
    samSht %>%  
        randomiseSampleTable(primaryGroup = primaryGroup)  %>%  
        addWells() %>%  
        randomiseWells()
}            


R/PlateLayoutPlots.R
# This script takes an experimental design table and generates a pseudo-block-randomised plate layout for it. It can cope with multiple plates - it will randomly spread the samples evenly between the plates. 
# The input table should be a csv file with a minimum of three columns headed "SampleName", SampleGroup" and "Replicate". An optional "Batch" column may also be specified. Any additional columns will be ignored.
# The outputs are one image file showing the plate layout, indicating both sample group and replicate number, and a table with the layout. The output table will include all the data in the input table, plus 3 or 4 additional columns indicating the plate number (if multiple plates), the row of the well (A-H), the column of the well (1-12) and the coordinate of the well (A1, A2...H7,H8)

library(optparse)

#Input options
options_list <- list(
  make_option(c('--layoutSheet', '-l'), type='character', help="Path to plate layout file - required", dest="layoutSheet", default="<undefined>"),
  make_option(c('--output', '-o'), type='character', help="Output Filename - optional", dest="outputFile", default="<undefined>"),
  make_option(c('--batchColumns', '-b'), type='character', help="Column to be plotted in multiplate experiments", dest="batCol", default="<undefined>")
)

#read options
parser <- OptionParser(option_list=options_list, usage="%prog [options]")
opts <- parse_args(parser)
layoutFile <- opts$layoutSheet
outputFile <- opts$outputFile
batchColumns <- opts$batCol

# set options that that have not beem provided
if(outputFile == "<undefined>") { outputFile <- gsub("[[:alnum:]]*$", "PlateLayoutPlots", basename(layoutFile)) }

######################################################################################################################################################################
# setwd("/run/user/1952417101/gvfs/sftp:host=clust1-headnode-1/home/sawle01/Groups/Ponder/20170109_OreillyM_PB_miRNAseq/PlateLayout/")
# layoutFile <- "miRNAseq.PlateLayout.csv"
# outputFile <- "TestRun1"
# batchColumns <- "Tissue,Outcome,Condition,SmokingStatus,Sex"


#########################################################################################################
## FUNCTIONS
#########################################################################################################

## Plate layout plot - shows the wells as a grid with sample groups represented by the colour and the replicate number printed in each well
testPlot <- function(plotDat, conditionToPlot, boxColCodes){
  textSize <- 0
  if(conditionToPlot=="SampleGroup") textSize <- 7
  ggplot(plotDat)+
    geom_tile(aes(x=Column, y=Row, fill=plotDat[[conditionToPlot]]), colour="black")  +
    geom_text(aes(x=plotDat$Column, y=plotDat$Row, label=ifelse(is.na(plotDat$Replicate), "", plotDat$Replicate)), size=textSize) +
    scale_x_continuous(breaks=1:NumberOfColumns) +
    scale_y_reverse(labels=LETTERS[1:8], breaks=c(1:8)) +
    scale_fill_manual(breaks=levels(plotDat[[conditionToPlot]]), values=boxColCodes) +
    theme(axis.text.x=element_text(size=20), axis.text.y=element_text(size=20),
          axis.title.x=element_blank(), axis.title.y=element_blank()) +
    ggtitle(conditionToPlot) +
    labs(fill=conditionToPlot)
}

## Plot multi-plate bar plot for assessing distribution of factors between the plates
#plotDat <- tabout
multiplatePlot <- function(plotDat, batches){
  
  #remove the genomics controls and water, and transform the plate numbers in to a factor
  plotDat <- plotDat %>%  
    filter(!SampleName%in%c("Water", "GenomicsControl")) %>% 
    mutate(PlateNumber=factor(PlateNumber))
  
  #plot the distribution of sample groups across plates
  sampleGroupPlot <- ggplot(plotDat)+
    geom_bar(aes(x=SampleGroup, fill=PlateNumber), position="dodge")+
    theme(axis.text.x=element_text(angle=90))+
    ggtitle("SampleGroup")
  print(sampleGroupPlot)
  
  #plot the distribution of replicate groups across plates
  replicatePlot <- ggplot(plotDat)+
    geom_bar(aes(x=Replicate, fill=PlateNumber), position="dodge")+
    theme(axis.text.x=element_text(angle=90))+
    ggtitle("Replicate")
  print(replicatePlot)
  
  #plot the distribution of other requested factors across plates
  if(batches!="<undefined>"){
    allbatches <- strsplit(batches, ",")[[1]]
    for(colName in allbatches){
      colNamePlot <- ggplot(plotDat, aes_string(x=colName, fill="PlateNumber"))+
        geom_bar(position="dodge")+
        theme(axis.text.x=element_text(angle=90))+
        ggtitle(colName)
      print(colNamePlot)
    }
  }
}
#########################################################################################################
## MAIN SCRIPT
#########################################################################################################

library(dplyr)
library(ggplot2)
library(RColorBrewer)


dat <- read.csv(layoutFile)
NumberOfPlates <- max(c(1, dat$PlateNumber))
groupsList <- levels(dat$SampleGroup)
groupsList <- groupsList[order(groupsList)]

#fix up the factors in the batch columns
if(batchColumns!="<undefined>"){
    allbatches <- strsplit(batchColumns, ",")[[1]]
    for(colName in allbatches){
        newLevels <- c(levels(dat[,colName])[!levels(dat[,colName])%in%c("Water", "GenomicsControl")], "Water", "GenomicsControl")
        dat[,colName] <- factor(dat[,colName], levels = newLevels)
    }
}
  
# randomise each plate
plateID <- ""
for(thisPlateNumber in 1:NumberOfPlates){
    
    if(NumberOfPlates>1){
        plateID <- paste(".Plate_", thisPlateNumber, sep="")
        #subset the sample data for this plate
        plateDat <- filter(dat, PlateNumber==thisPlateNumber)
    }else{
        plateDat <- dat
    }

    #replace the well row letters with numbers
    plateDat <- plateDat %>% 
        mutate(Row=match(Row, LETTERS))
    
    NumberOfColumns <- max(plateDat$Column)
    nWater <- length(which(plateDat$SampleName=="Water"))
    nGenomicControls <- length(which(plateDat$SampleName=="GenomicsControl"))
    
    # plot the plate layout and save
     #generate a colour coding for plotting to keep consistency across different plates
    if(thisPlateNumber==1){
        colCodes <- colorRampPalette(brewer.pal(12,"Set3"))(length(levels(plateDat$SampleGroup))-2)
        if(nWater>0) colCodes <- c(colCodes, "#0000FF")
        if(nGenomicControls>0) colCodes <- c(colCodes, "#000000")
    }
    #plot plate layout
    testPlot(plateDat, "SampleGroup", colCodes)
    plotWidth <- (1.1*NumberOfColumns)+(max(nchar(groupsList))*0.075)+0.15
    ggsave(paste(outputFile, plateID, ".png", sep=""), height = 5, width=plotWidth)  
    
    if(batchColumns!="<undefined>"){
      allbatches <- strsplit(batchColumns, ",")[[1]]
      #generate a colour coding for plotting each batch to keep consistency across different plates
      if(thisPlateNumber==1){ 
        colCodesBatches <- vector("list", length(allbatches)) 
        names(colCodesBatches) <- allbatches
        }
      for(colName in allbatches){
        if(thisPlateNumber==1){
          tempCols <- colorRampPalette(brewer.pal(12,"Set3"))(length(levels(plateDat[[colName]]))-2)
          if(nWater>0) tempCols <- c(tempCols, "#0000FF")
          if(nGenomicControls>0) tempCols <- c(tempCols, "#000000")
          colCodesBatches[[colName]] <- tempCols
        }
        #plot
        testPlot(plateDat, colName, colCodesBatches[[colName]])
        plotWidth <- (1.1*NumberOfColumns)+1.5
        ggsave(paste(outputFile, plateID, ".", colName, ".png", sep=""), height = 5, width=plotWidth)  
      }
    }
}

#create a pdf with plots showing the distribution of the different factors across plates
if(NumberOfPlates>1){
    pdf(paste(outputFile, ".Multiplate_Distribution_Check.pdf", sep=""))
    multiplatePlot(dat[dat$SampleName!="GenomicsControl",], batchColumns)
    dev.off()
}

R/MainFunctions.R
#' Randomize Single Plate
#'
#' Randomize up to 96 samples across a single 96 well plate. 
#' 
#' @param designTable The meta data table - a data.frame/tibble.
#' @param outputFile The root name for the output files if required.
#' @param primaryGroup The name of a column in the metadata table to use for
#' optimising the distribution of samples across the plate
#' @param batchColumns A character vector wit the names of additional columns
#' in the metadata table to use in assessing the distribution of the samples
#' across the plate.
#' @param nIter The number of random layouts to generate and assess.
#' @param nCores The number of cores to use. Set to 1 to run serial.
#' @details The metadata (\code{designTable}) should be a data.frame/tibble. The 
#' column names can be anything. By default the function expects one column 
#' named \code{SampleGroup} that it will use to distribute the samples across
#' the plate. This can be set to any other column using the \code{primaryGroup}
#' argument.
#' @details The columns specified in \code{batchColumns} are used in addition
#' to \code{primaryGroup} to assess which of the \code{nIter} layouts is 
#' optimal in terms of the distribution of samples. This is assessed based on
#' the sum of Euclidean distances between replicate samples within each column.
#' @return If no \code{outputFile} is provided, a tibble containing the original
#' metadata and the well assignments for each sample. If an \code{outputFile} is
#' provided, nothing - the function writes out the plate layout to a tsv file 
#' and a plot for each of `primaryGroup` and `batchColumns` to png files.
#' @examples
#' library(readr)
#' designSheet <- system.file("extdata", "metadata_12x3.tsv",
#'                            package = "PlateLayout")
#' bColumns <- c("ExtractionInformation", "PassageNumber") 
#' designTable <- read_tsv(designSheet)
#' # Create an R object in the session
#' plateLayout <- randomizeSinglePlate(designTable,  batchColumns = bColumns,
#'                      nIter = 100) 
#'
#' # Or directly output a table and plots
#' outputFile <- "Test"
#' randomizeSinglePlate(designTable, 
#'                      outputFile = outputFile, 
#'                      batchColumns = bColumns,
#'                      nIter = 100) 
#' @export
randomizeSinglePlate <- function(designTable, 
                                 outputFile =  NULL, 
                                 primaryGroup = "SampleGroup", 
                                 batchColumns = NULL, 
                                 nIter = 10000,
                                 nCores = 4){
    # Set parallel processing
    options(future.supportsMulticore.unstable="quiet") # turn off warning about forking in RStudio
    if(nCores>1){ plan(multicore, workers = nCores) }

    # Checks
    if(length(primaryGroup) > 1) {
        stop("Only specify 1 column for the 'primaryGroup'")
    }
    if(!primaryGroup%in%colnames(designTable)){
        stop("'", primaryGroup, "' is not a column in the design sheet")
    }
    if(any(!batchColumns%in%colnames(designTable))){
       stop("Not all of the provided 'batchColumns' are in the design sheet")
    }
    if(nrow(designTable) > 96){
        stop("There are too many samples in the sample sheet for a single plate")
    }
    if(nrow(designTable) > 94){
        warning("There are more than 94 samples. Are you sure you do not ",
                "want to leave wells for the Genomics positive controls?")
    }

    # Modify batch columns
    bCols <- c(primaryGroup, batchColumns) %>%  
        str_subset("Replicate", negate = TRUE) %>%  
        unique()

    message("Run ", nIter, " random layouts")
    options(future.rng.onMisuse = "ignore")
    layouts <- future_map(1:nIter, 
                          ~plateRandomisation(designTable, 
                                              primaryGroup = primaryGroup),
                          .progress = TRUE,
                          seed = TRUE)

    message("Calculate distribution scores for each layout and find best layout")
    bestLayout <- getMinScore(layouts, bCols)
        
    # Choose best layout
    finalLayout <- pluck(layouts, bestLayout)

    if(!is.null(outputFile)){
        message("Output plots and layout")
        # Save a plot for each batch column
        bCols %>%  
            walk(outputPlatePlot, datTab = finalLayout, outFileName = outputFile)    

        # Write out the layout
        finalLayout %>%  
            select(-RowID) %>%  
            write_tsv(str_c(outputFile, ".PlateLayout.tsv"))
    }else{
        return(finalLayout)
    }
}
